### PR TITLE
Add function stdlib::sort_by

### PR DESCRIPTION
**Pull request by XMol**
## Summary
Add new Ruby function `stdlib::sort_by()`, emulating native Ruby's `Enumerable.sort_by()` for Puppet.

## Additional Context
`stdlib::sort_by()` supports sorting Strings, Arrays and Hashes
- Arrays, since that is the most prominent use-case
- Strings, because it is a subset of Iterable, even though I cannot imagine any sensible proc besides `clone`/`copy`, which would be identical to basic `String::sort()`
- Hashes, because that was found to be missing in #1376, even though there likely is little demand for it

## Related Issues
#1376

## Checklist
- [x] 🟢 Spec tests included
- [ ] ⚪Acceptance tests are not changed, some of them fail even on the master branch, so unrelated to my additions
- [x] Manually verified with `puppet apply` version 7.24.0 on RHEL-7.9

### CODE REVIEW

File: lib/puppet/functions/stdlib/sort_by.rb
Comment (smortex) [inline lib/puppet/functions/stdlib/sort_by.rb:31]: I guess we should document that with this form, the block get passed an array of 2 items, the first being the key, and the second being the value?

File: lib/puppet/functions/stdlib/sort_by.rb
Comment (smortex) [inline lib/puppet/functions/stdlib/sort_by.rb:39]: Same here, we should document how it is different from the previous form by accepting a block that receive the key and the value as 2 arguments.

File: lib/puppet/functions/stdlib/sort_by.rb
Comment (smortex) [inline lib/puppet/functions/stdlib/sort_by.rb:58]: `e` makes me think, I think this is more readable with explicit names:
```suggestion
    sort_by_iterable(hsh) { |key, value| yield(key, value) }.to_h
```

File: lib/puppet/functions/stdlib/sort_by.rb
Comment (smortex) [inline lib/puppet/functions/stdlib/sort_by.rb:54]: This `to_h` makes `sort_by` behave differently that Ruby's `sort_by` which return an array when sorting a Hash (you have to manually `Hash(...)` the result if you want a hash).

I wonder if we should match Ruby or not.  As a Puppet and Ruby developer, I trend to expect some consistency, but obviously the Puppet language is not Ruby so who am I to expect this :shrug:?  Maybe some people have an idea about this?

File: lib/puppet/functions/stdlib/sort_by.rb
Comment (XMol) [reply lib/puppet/functions/stdlib/sort_by.rb:58]: `sort_by_hash2` is written this way, because Ruby's `sort_by` expects a block that has just one parameter. I thought, this construct was needed to align `stdlib::sort_by($hash) |$key, $value| { ... }` with that. Though it looks like I was mistaken and Ruby handles that transparently:
```ruby
hsh = { 'The' => 'quick', 'brown' => 'fox', 'jumps' => 'over', 'the' => 'lazy', 'dog' => '.' }
hsh.sort_by { |e| e[1] } # => [["dog", "."], ["brown", "fox"], ["the", "lazy"], ["jumps", "over"], ["The", "quick"]]
hsh.sort_by { |k, v| v } # => [["dog", "."], ["brown", "fox"], ["the", "lazy"], ["jumps", "over"], ["The", "quick"]]
```
My delusion was reinforced by `sort_by()` accepting blocks with arity 2 on Arrays and _not_ interpreting that as index-element-tuple, but rather element-nil-tuple. In other words, `sort_by()` does _not_ use `Enumerable.each_with_index` when appropriate.

Comment (XMol) [reply lib/puppet/functions/stdlib/sort_by.rb:58]: On that note, [Puppet's `each()`](https://github.com/puppetlabs/puppet/blob/8.0.1_release/lib/puppet/functions/each.rb#L153) explicitly calls Ruby's `each_with_index()` in this case.

File: lib/puppet/functions/stdlib/sort_by.rb
Comment (XMol) [reply lib/puppet/functions/stdlib/sort_by.rb:54]: Keeping consistency means that we can simplify the code enormously: only `sort_by_iterable()` is needed, if Ruby can handle blocks with two or a single parameter transparently for us. I went out of my way to distinct between Strings, Arrays and Hashes, precisely because _my_ expectation was that `sort_by()` should return the same data structure that it was fed. If we cannot find an objective argument for either interpretation, we will be unable to decide the "correct" resolution, I guess.

File: lib/puppet/functions/stdlib/sort_by.rb
Comment (XMol) [reply lib/puppet/functions/stdlib/sort_by.rb]: With your other hints I decided to collapse `sort_by_hash1` and `sort_by_hash2`, plus explicitly stating that single and double parameter blocks are accepted.

File: lib/puppet/functions/stdlib/sort_by.rb
Comment (XMol) [reply lib/puppet/functions/stdlib/sort_by.rb]: With the collapsed `sort_by_hash()`, this discussion can be resolved, too.

File: lib/puppet/functions/stdlib/sort_by.rb
Comment (smortex) [reply lib/puppet/functions/stdlib/sort_by.rb:47]: > _my_ expectation was that `sort_by()` should return the same data structure that it was fed.

Good point. I think that help: Puppet can currently sort a String and return a String, something Ruby cannot do because `String#sort` does not exist.  So I am fine with returning a Hash when given a Hash because it makes more sense IMHO!

File: lib/puppet/functions/stdlib/sort_by.rb
Comment (smortex) [inline lib/puppet/functions/stdlib/sort_by.rb:31]: Is this different than:

```suggestion
    block_param 'Callable[1,2]', :block
```

**Review by smortex — APPROVED** [review]
This looks good, I am curious what other people think about it.

**Comment by XMol** [comment]
Regarding the pending signature of the CLA agreement, I'm still waiting for a reply from our legal department... No estimated timeline, of course.

**Comment by XMol** [comment]
Since our legal department has not replied to me in 3 months, I sign the CLA at my own risk. Stalling this MR for an unknown amount of time just because I'm trying to follow the rules is ridiculous.

**Comment by XMol** [comment]
Hi project maintainers,

is anything else missing to merge this PR? I've signed the CLA months ago and at least one reviewer gave his approval (thanks for the support, @smortex).

Kind regards,  
Xavier.

**Comment by bastelfreak** [comment]
@XMol can you please rebase against the main branch?

**Comment by XMol** [comment]
> @XMol can you please rebase against the main branch?

Will do...

**Comment by bastelfreak** [comment]
As I wrote, please do a *rebase*, not a merge.

**Comment by XMol** [comment]
> As I wrote, please do a _rebase_, not a merge.

Sorry, I understood that, but I wasn't sure what GitHub would do when I press "sync fork". :-)

**Comment by bastelfreak** [comment]
yeah the button is shit :(

**Comment by XMol** [comment]
Guess we now need to wait for the two pipelines to eventually fail in about 4 hours? The Acceptance tests on Ubuntu-ARM seem to be broken...

**Review by bastelfreak — APPROVED** [review]

**Comment by XMol** [comment]
Thanks for the prompt reaction, @bastelfreak.

**Pull request merged**
